### PR TITLE
Made `Ref`, `SynchronizedRef`,`RcRef, and `SubscriptionRef` a subtype of `Effect`

### PR DESCRIPTION
**Pull request by KhraksMamtsov**
## Type

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Related

- Related Issue https://github.com/Effect-TS/effect/issues/3500
- Closes #

### CODE REVIEW

**Comment by tim-smart** [comment]
I wonder if making `Readable` a sub-type of Effect is more appropriate.

**Comment by KhraksMamtsov** [comment]
If I make `Readable` a subtype of `Effect` this error arise:
image
Readable looks more like mixin or trait - and only finalized types should be Effect's subtypes

Also where is problem with `Unify` - now it is designed to work with covariant types
But since `*Ref` has an invariant type parameter, this does not work
```ts
type Test = Unify<Ref<1> | Ref<"a">> // Ref<1> | Ref<"a"> | Ref<1 | "a">
```
because `Unify` uses ` ... | Exclude<X, Z> | ...` and `Exclude<Ref<in out 1> | Ref<in out 'a'>, Ref<in out 1 | 'a'>` takes no effect

**Comment by KhraksMamtsov** [comment]
I figured out with Unify module - all invariant types must remain as they are
add dtslint tests

**Comment by KhraksMamtsov** [comment]
@tim-smart Could you please tell me if this is a move in the right direction? or is the correct implementation much more complicated than this approach? Thank you

**Comment by tim-smart** [comment]
I'm still not sure if I like having `Ref`'s a subtype of Effect, it feels a bit strange. If others like it I'm happy to be veto'ed though.

**Comment by mikearnaldi** [comment]
> I'm still not sure if I like having `Ref`'s a subtype of Effect, it feels a bit strange. If others like it I'm happy to be veto'ed though.

kind of goes in the same direction of having Option be an Effect

**Comment by mikearnaldi** [comment]
I suppose we can accept for now and if we don't like it revise in 4.0

**Comment by KhraksMamtsov** [comment]
> kind of goes in the same direction of having Option be an Effect

yes, after turning Option and Either into an Effect, the other subtypes don't look so wild, but even more ergonomic. It seems to be the next logical step.

File: packages/effect/src/internal/subscriptionRef.ts [hunk 1]
```diff
@@ -37,17 +37,18 @@ class SubscriptionRefImpl<in out A> implements SubscriptionRef.SubscriptionRef<A
     readonly pubsub: PubSub.PubSub<A>,
     readonly semaphore: Effect.Semaphore
   ) {
+    super()
     this[Readable.TypeId] = Readable.TypeId
     this[Subscribable.TypeId] = Subscribable.TypeId
-    this.get = Ref.get(this.ref)
```
Comment (tim-smart) [inline packages/effect/src/internal/subscriptionRef.ts:42]: I think internally we should still use the module APIs.

**Review by tim-smart — CHANGES_REQUESTED** [review]

File: packages/effect/src/internal/subscriptionRef.ts (hunk 1 above)
Comment (KhraksMamtsov) [reply packages/effect/src/internal/subscriptionRef.ts:42]: Done

**Review by tim-smart — APPROVED** [review]

**Pull request merged**